src/aws-cpp-sdk-core/source/http/crt/CRTHttpClient.cpp

```diff
@@ -0,0 +1,639 @@
+            };
+
+            std::shared_ptr<Crt::Http::HttpClientConnection> connectionRef(nullptr);
+
+            // now we finally have the request, get a connection and make the request.
+            connectionManager->AcquireConnection(
+                    [&connectionRef, &requestOptions, response, &waiter, request, this]
+                    (std::shared_ptr<Crt::Http::HttpClientConnection> connection, int errorCode)
+                    {
+                        OnClientConnectionAvailable(connection, errorCode, connectionRef, requestOptions, waiter, request, response, *this);
+                    });
+
+            bool waiterTimedOut = false;
+            // Naive http request timeout implementation. This doesn't factor in how long it took to get the connection from the pool, and
+            // I'm undecided on the queueing theory implications of this decision so if this turns out to be the wrong granularity
+            // this is the section of code you should be changing. You can probably get "close" by having an additional
+            // atomic (not necessarily full on atomics implementation, but it needs to be the size of a WORD if it's not)
+            // counter that gets incremented in the acquireConnection callback as long as your connection timeout
+            // is shorter than your request timeout. Even if it's not, that would handle like.... 4-5 nines of getting this right.
+            // since in the worst case scenario, your connect timeout got preempted by the request timeout, and is it really worth
+            // all that effort if that's the worst thing that can happen?
+            if (m_configuration.requestTimeoutMs > 0 )
+            {
+                auto requestExpiryTime = std::chrono::high_resolution_clock::now() +
+                                         std::chrono::milliseconds(m_configuration.requestTimeoutMs);
+                waiterTimedOut = !waiter.WaitOnCompletionUntil(requestExpiryTime);
+
+                // if this is true, the waiter timed out without a terminal condition being woken up.
+                if (waiterTimedOut)
+                {
+                    // close the connection if it's still there so we can expedite anything we're waiting on.
+                    if (connectionRef)
+                    {
+                        connectionRef->Close();
+                    }
+                }
+            }
+
+            // always wait, even if the above section timed out, because Wakeup() hasn't yet been called,
+            // and this means we're still waiting on some queued up callbacks to fire.
+            // going past this point before that occurs will cause a segfault when the callback DOES finally fire
+            // since the waiter is on the stack.
+            waiter.WaitOnCompletion();
+
+            // now handle if we timed out or not.
+            if (waiterTimedOut)
+            {
+                response->SetClientErrorType(

```

Comment:
debatable: don't set the REQUEST_TIMEOUT error if the response actually succeeded.

there's a decent chance you tried to kill it but it still succeeded